File: backend/feature/card/src/main/java/com/veda/server/controller/CategoryController.java
package com.veda.server.controller;

import com.veda.server.dto.request.CategoryRequest;
import com.veda.server.dto.request.UpdateCategoryRequest;
import com.veda.server.dto.response.CategoryResponse;
import com.veda.server.service.CategoryService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/cards/categories")
public class CategoryController {

    private final CategoryService categoryService;

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public ResponseEntity<CategoryResponse> createCategory(@Valid @RequestBody CategoryRequest request) {
        return ResponseEntity.ok(categoryService.createCategory(request));
    }

    @GetMapping
    public ResponseEntity<List<CategoryResponse>> getAllCategories(@RequestParam(defaultValue = "10") Integer limit) {
        return ResponseEntity.ok(categoryService.getAllCategories(limit));
    }

    @GetMapping("/{id}")
    public ResponseEntity<CategoryResponse> getCategoryById(@PathVariable Integer id) {
        return ResponseEntity.ok(categoryService.getCategoryById(id));
    }

    @PutMapping("/{id}")
    public ResponseEntity<CategoryResponse> updateCategory(
            @PathVariable Integer id,
            @Valid @RequestBody UpdateCategoryRequest request
    ) {
        return ResponseEntity.ok(categoryService.updateCategory(id, request));
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteCategoryById(@PathVariable Integer id) {
        categoryService.deleteCategoryById(id);
    }
}

File: feature/admin/src/main/kotlin/com/titsuko/controller/CategoryController.kt
package com.titsuko.controller

import com.titsuko.dto.request.CategoryRequest
import com.titsuko.service.CategoryService
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*

@Controller
@RequestMapping("/admin/category")
class CategoryController(
    private val categoryService: CategoryService
) {

    @GetMapping
    fun getPage(model: Model): String {
        val categories = categoryService.getAllCategories()
        model.addAttribute("categories", categories)
        return "category"
    }

    @PostMapping("/create")
    fun createCategory(@ModelAttribute request: CategoryRequest): String {
        try {
            categoryService.createCategory(request)
        } catch (e: Exception) {
            return "redirect:/admin/category?error=${e.message}"
        }
        return "redirect:/admin/category"
    }

    @PostMapping("/{id}/update")
    fun updateCategory(@PathVariable id: Long, @ModelAttribute request: CategoryRequest): String {
        try {
            categoryService.updateCategory(id, request)
        } catch (e: Exception) {
            return "redirect:/admin/category?error=${e.message}"
        }
        return "redirect:/admin/category"
    }

    @PostMapping("/{id}/delete")
    fun deleteCategory(@PathVariable id: Long): String {
        try {
            categoryService.deleteCategory(id)
        } catch (e: Exception) {
            return "redirect:/admin/category?error=${e.message}"
        }
        return "redirect:/admin/category"
    }
}